share/man/man4/tcp.4
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.
.\" Copyright (c) 2010-2011 The FreeBSD Foundation
.\" All rights reserved.
.\"
.\" Portions of this documentation were written at the Centre for Advanced
.\" Internet Architectures, Swinburne University of Technology, Melbourne,
.\" Australia by David Hayes under sponsorship from the FreeBSD Foundation.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     From: @(#)tcp.4	8.1 (Berkeley) 6/5/93
.\" $FreeBSD: releng/12.0/share/man/man4/tcp.4 337390 2018-08-06 18:42:37Z jtl $
.\"
.Dd August 6, 2018
.Dt TCP 4
.Os
.Sh NAME
.Nm tcp
.Nd Internet Transmission Control Protocol
.Sh SYNOPSIS
.In sys/types.h
.In sys/socket.h
.In netinet/in.h
.In netinet/tcp.h
.Ft int
.Fn socket AF_INET SOCK_STREAM 0
.Sh DESCRIPTION
The
.Tn TCP
protocol provides reliable, flow-controlled, two-way
transmission of data.
It is a byte-stream protocol used to
support the
.Dv SOCK_STREAM
abstraction.
.Tn TCP
uses the standard
Internet address format and, in addition, provides a per-host
collection of
.Dq "port addresses" .
Thus, each address is composed
of an Internet address specifying the host and network,
with a specific
.Tn TCP
port on the host identifying the peer entity.
.Pp
Sockets utilizing the
.Tn TCP
protocol are either
.Dq active
or
.Dq passive .
Active sockets initiate connections to passive
sockets.
By default,
.Tn TCP
sockets are created active; to create a
passive socket, the
.Xr listen 2
system call must be used
after binding the socket with the
.Xr bind 2
system call.
Only passive sockets may use the
.Xr accept 2
call to accept incoming connections.
Only active sockets may use the
.Xr connect 2
call to initiate connections.
.Pp
Passive sockets may
.Dq underspecify
their location to match
incoming connection requests from multiple networks.
This technique, termed
.Dq "wildcard addressing" ,
allows a single
server to provide service to clients on multiple networks.
To create a socket which listens on all networks, the Internet
address
.Dv INADDR_ANY
must be bound.
The
.Tn TCP
port may still be specified
at this time; if the port is not specified, the system will assign one.
Once a connection has been established, the socket's address is
fixed by the peer entity's location.
The address assigned to the
socket is the address associated with the network interface
through which packets are being transmitted and received.
Normally, this address corresponds to the peer entity's network.
.Pp
.Tn TCP
supports a number of socket options which can be set with
.Xr setsockopt 2
and tested with
.Xr getsockopt 2 :
.Bl -tag -width ".Dv TCP_FUNCTION_BLK"
.It Dv TCP_INFO
Information about a socket's underlying TCP session may be retrieved
by passing the read-only option
.Dv TCP_INFO
to
.Xr getsockopt 2 .
It accepts a single argument: a pointer to an instance of
.Vt "struct tcp_info" .
.Pp
This API is subject to change; consult the source to determine
which fields are currently filled out by this option.
.Fx
specific additions include
send window size,
receive window size,
and
bandwidth-controlled window space.
.It Dv TCP_CCALGOOPT
Set or query congestion control algorithm specific parameters.
See
.Xr mod_cc 4
for details.
.It Dv TCP_CONGESTION
Select or query the congestion control algorithm that TCP will use for the
connection.
See
.Xr mod_cc 4
for details.
.It Dv TCP_FUNCTION_BLK
Select or query the set of functions that TCP will use for this connection.
This allows a user to select an alternate TCP stack.
The alternate TCP stack must already be loaded in the kernel.
To list the available TCP stacks, see
.Va functions_available
in the
.Sx MIB Variables
section further down.
To list the default TCP stack, see
.Va functions_default
in the
.Sx MIB Variables
section.
.It Dv TCP_KEEPINIT
This
.Xr setsockopt 2
option accepts a per-socket timeout argument of
.Vt "u_int"
in seconds, for new, non-established
.Tn TCP
connections.
For the global default in milliseconds see
.Va keepinit
in the
.Sx MIB Variables
section further down.
.It Dv TCP_KEEPIDLE
This
.Xr setsockopt 2
option accepts an argument of
.Vt "u_int"
for the amount of time, in seconds, that the connection must be idle
before keepalive probes (if enabled) are sent for the connection of this
socket.
If set on a listening socket, the value is inherited by the newly created
socket upon
.Xr accept 2 .
For the global default in milliseconds see
.Va keepidle
in the
.Sx MIB Variables
section further down.
.It Dv TCP_KEEPINTVL
This
.Xr setsockopt 2
option accepts an argument of
.Vt "u_int"
to set the per-socket interval, in seconds, between keepalive probes sent
to a peer.
If set on a listening socket, the value is inherited by the newly created
socket upon
.Xr accept 2 .
For the global default in milliseconds see
.Va keepintvl
in the
.Sx MIB Variables
section further down.
.It Dv TCP_KEEPCNT
This
.Xr setsockopt 2
option accepts an argument of
.Vt "u_int"
and allows a per-socket tuning of the number of probes sent, with no response,
before the connection will be dropped.
If set on a listening socket, the value is inherited by the newly created
socket upon
.Xr accept 2 .
For the global default see the
.Va keepcnt
in the
.Sx MIB Variables
section further down.
.It Dv TCP_NODELAY
Under most circumstances,
.Tn TCP
sends data when it is presented;
when outstanding data has not yet been acknowledged, it gathers
small amounts of output to be sent in a single packet once
an acknowledgement is received.
For a small number of clients, such as window systems
that send a stream of mouse events which receive no replies,
this packetization may cause significant delays.
The boolean option
.Dv TCP_NODELAY
defeats this algorithm.
.It Dv TCP_MAXSEG
By default, a sender- and
.No receiver- Ns Tn TCP
will negotiate among themselves to determine the maximum segment size
to be used for each connection.
The
.Dv TCP_MAXSEG
option allows the user to determine the result of this negotiation,
and to reduce it if desired.
.It Dv TCP_NOOPT
.Tn TCP
usually sends a number of options in each packet, corresponding to
various
.Tn TCP
extensions which are provided in this implementation.
The boolean option
.Dv TCP_NOOPT
is provided to disable
.Tn TCP
option use on a per-connection basis.
.It Dv TCP_NOPUSH
By convention, the
.No sender- Ns Tn TCP
will set the
.Dq push
bit, and begin transmission immediately (if permitted) at the end of
every user call to
.Xr write 2
or
.Xr writev 2 .
When this option is set to a non-zero value,
.Tn TCP
will delay sending any data at all until either the socket is closed,
or the internal send buffer is filled.
.It Dv TCP_MD5SIG
This option enables the use of MD5 digests (also known as TCP-MD5)
on writes to the specified socket.
Outgoing traffic is digested;
digests on incoming traffic are verified.
When this option is enabled on a socket, all inbound and outgoing
TCP segments must be signed with MD5 digests.
.Pp
One common use for this in a
.Fx
router deployment is to enable
based routers to interwork with Cisco equipment at peering points.
Support for this feature conforms to RFC 2385.
.Pp
In order for this option to function correctly, it is necessary for the
administrator to add a tcp-md5 key entry to the system's security
associations database (SADB) using the
.Xr setkey 8
utility.
This entry can only be specified on a per-host basis at this time.
.Pp
If an SADB entry cannot be found for the destination,
the system does not send any outgoing segments and drops any inbound segments.
.Pp
Each dropped segment is taken into account in the TCP protocol statistics.
.El
.Pp
The option level for the
.Xr setsockopt 2
call is the protocol number for
.Tn TCP ,
available from
.Xr getprotobyname 3 ,
or
.Dv IPPROTO_TCP .
All options are declared in
.In netinet/tcp.h .
.Pp
Options at the
.Tn IP
transport level may be used with
.Tn TCP ;
see
.Xr ip 4 .
Incoming connection requests that are source-routed are noted,
and the reverse source route is used in responding.
.Pp
The default congestion control algorithm for
.Tn TCP
is
.Xr cc_newreno 4 .
Other congestion control algorithms can be made available using the
.Xr mod_cc 4
framework.
.Ss MIB Variables
The
.Tn TCP
protocol implements a number of variables in the
.Va net.inet.tcp
branch of the
.Xr sysctl 3
MIB.
.Bl -tag -width ".Va TCPCTL_DO_RFC1323"
.It Dv TCPCTL_DO_RFC1323
.Pq Va rfc1323
Implement the window scaling and timestamp options of RFC 1323
(default is true).
.It Dv TCPCTL_MSSDFLT
.Pq Va mssdflt
The default value used for the maximum segment size
.Pq Dq MSS
when no advice to the contrary is received from MSS negotiation.
.It Dv TCPCTL_SENDSPACE
.Pq Va sendspace
Maximum
.Tn TCP
send window.
.It Dv TCPCTL_RECVSPACE
.Pq Va recvspace
Maximum
.Tn TCP
receive window.
.It Va log_in_vain
Log any connection attempts to ports where there is not a socket
accepting connections.
The value of 1 limits the logging to
.Tn SYN
(connection establishment) packets only.
That of 2 results in any
.Tn TCP
packets to closed ports being logged.
Any value unlisted above disables the logging
(default is 0, i.e., the logging is disabled).
.It Va msl
The Maximum Segment Lifetime, in milliseconds, for a packet.
.It Va keepinit
Timeout, in milliseconds, for new, non-established
.Tn TCP
connections.
The default is 75000 msec.
.It Va keepidle
Amount of time, in milliseconds, that the connection must be idle
before keepalive probes (if enabled) are sent.
The default is 7200000 msec (2 hours).
.It Va keepintvl
The interval, in milliseconds, between keepalive probes sent to remote
machines, when no response is received on a
.Va keepidle
probe.
The default is 75000 msec.
.It Va keepcnt
Number of probes sent, with no response, before a connection
is dropped.
The default is 8 packets.
.It Va always_keepalive
Assume that
.Dv SO_KEEPALIVE
is set on all
.Tn TCP
connections, the kernel will
periodically send a packet to the remote host to verify the connection
is still up.
.It Va icmp_may_rst
Certain
.Tn ICMP
unreachable messages may abort connections in
.Tn SYN-SENT
state.
.It Va do_tcpdrain
Flush packets in the
.Tn TCP
reassembly queue if the system is low on mbufs.
.It Va blackhole
If enabled, disable sending of RST when a connection is attempted
to a port where there is not a socket accepting connections.
See
.Xr blackhole 4 .
.It Va delayed_ack
Delay ACK to try and piggyback it onto a data packet.
.It Va delacktime
Maximum amount of time, in milliseconds, before a delayed ACK is sent.
.It Va path_mtu_discovery
Enable Path MTU Discovery.
.It Va tcbhashsize
Size of the
.Tn TCP
control-block hash table
(read-only).
This may be tuned using the kernel option
.Dv TCBHASHSIZE
or by setting
.Va net.inet.tcp.tcbhashsize
in the
.Xr loader 8 .
.It Va pcbcount
Number of active process control blocks
(read-only).
.It Va syncookies
Determines whether or not
.Tn SYN
cookies should be generated for outbound
.Tn SYN-ACK
packets.
.Tn SYN
cookies are a great help during
.Tn SYN
flood attacks, and are enabled by default.
(See
.Xr syncookies 4 . )
.It Va isn_reseed_interval
The interval (in seconds) specifying how often the secret data used in
RFC 1948 initial sequence number calculations should be reseeded.
By default, this variable is set to zero, indicating that

cddl/lib/libdtrace/tcp.d
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 *
 * $FreeBSD: releng/12.0/cddl/lib/libdtrace/tcp.d 336285 2018-07-14 17:12:04Z tuexen $
 */
/*
 * Copyright (c) 2010, Oracle and/or its affiliates. All rights reserved.
 * Copyright (c) 2013 Mark Johnston <markj@freebsd.org>
 */

#pragma D depends_on library ip.d
#pragma D depends_on module kernel
#pragma D depends_on provider tcp

/*
 * Convert a TCP state value to a string.
 */
#pragma D binding "1.6.3" TCPS_CLOSED
inline int TCPS_CLOSED =	0;
#pragma D binding "1.6.3" TCPS_LISTEN
inline int TCPS_LISTEN =	1;
#pragma D binding "1.6.3" TCPS_SYN_SENT
inline int TCPS_SYN_SENT =	2;
#pragma D binding "1.6.3" TCPS_SYN_RECEIVED
inline int TCPS_SYN_RECEIVED =	3;
#pragma D binding "1.6.3" TCPS_ESTABLISHED
inline int TCPS_ESTABLISHED =	4;
#pragma D binding "1.6.3" TCPS_CLOSE_WAIT
inline int TCPS_CLOSE_WAIT =	5;
#pragma D binding "1.6.3" TCPS_FIN_WAIT_1
inline int TCPS_FIN_WAIT_1 =	6;
#pragma D binding "1.6.3" TCPS_CLOSING
inline int TCPS_CLOSING =	7;
#pragma D binding "1.6.3" TCPS_LAST_ACK
inline int TCPS_LAST_ACK =	8;
#pragma D binding "1.6.3" TCPS_FIN_WAIT_2
inline int TCPS_FIN_WAIT_2 =	9;
#pragma D binding "1.6.3" TCPS_TIME_WAIT
inline int TCPS_TIME_WAIT =	10;

/*
 * For compatibility also provide the names used by Solaris.
 */
#pragma D binding "1.13" TCP_STATE_CLOSED
inline int TCP_STATE_CLOSED =		TCPS_CLOSED;
#pragma D binding "1.13" TCP_STATE_LISTEN
inline int TCP_STATE_LISTEN =		TCPS_LISTEN;
#pragma D binding "1.13" TCP_STATE_SYN_SENT
inline int TCP_STATE_SYN_SENT =		TCPS_SYN_SENT;
#pragma D binding "1.13" TCP_STATE_SYN_RECEIVED
inline int TCP_STATE_SYN_RECEIVED =	TCPS_SYN_RECEIVED;
#pragma D binding "1.13" TCP_STATE_ESTABLISHED
inline int TCP_STATE_ESTABLISHED =	TCPS_ESTABLISHED;
#pragma D binding "1.13" TCP_STATE_CLOSE_WAIT
inline int TCP_STATE_CLOSE_WAIT =	TCPS_CLOSE_WAIT;
#pragma D binding "1.13" TCP_STATE_FIN_WAIT_1
inline int TCP_STATE_FIN_WAIT_1 =	TCPS_FIN_WAIT_1;
#pragma D binding "1.13" TCP_STATE_CLOSING
inline int TCP_STATE_CLOSING =		TCPS_CLOSING;
#pragma D binding "1.13" TCP_STATE_LAST_ACK
inline int TCP_STATE_LAST_ACK =		TCPS_LAST_ACK;
#pragma D binding "1.13" TCP_STATE_FIN_WAIT_2
inline int TCP_STATE_FIN_WAIT_2 =	TCPS_FIN_WAIT_2;
#pragma D binding "1.13" TCP_STATE_TIME_WAIT
inline int TCP_STATE_TIME_WAIT =	TCPS_TIME_WAIT;

/* TCP segment flags. */
#pragma D binding "1.6.3" TH_FIN
inline uint8_t TH_FIN =		0x01;
#pragma D binding "1.6.3" TH_SYN
inline uint8_t TH_SYN =		0x02;
#pragma D binding "1.6.3" TH_RST
inline uint8_t TH_RST =		0x04;
#pragma D binding "1.6.3" TH_PUSH
inline uint8_t TH_PUSH =	0x08;
#pragma D binding "1.6.3" TH_ACK
inline uint8_t TH_ACK =		0x10;
#pragma D binding "1.6.3" TH_URG
inline uint8_t TH_URG =		0x20;
#pragma D binding "1.6.3" TH_ECE
inline uint8_t TH_ECE =		0x40;
#pragma D binding "1.6.3" TH_CWR
inline uint8_t TH_CWR =		0x80;

/* TCP connection state strings. */
#pragma D binding "1.6.3" tcp_state_string
inline string tcp_state_string[int32_t state] =
	state == TCPS_CLOSED ?		"state-closed" :
	state == TCPS_LISTEN ?		"state-listen" :
	state == TCPS_SYN_SENT ?	"state-syn-sent" :
	state == TCPS_SYN_RECEIVED ?	"state-syn-received" :
	state == TCPS_ESTABLISHED ?	"state-established" :
	state == TCPS_CLOSE_WAIT ?	"state-close-wait" :
	state == TCPS_FIN_WAIT_1 ?	"state-fin-wait-1" :
	state == TCPS_CLOSING ?		"state-closing" :
	state == TCPS_LAST_ACK ?	"state-last-ack" :
	state == TCPS_FIN_WAIT_2 ?	"state-fin-wait-2" :
	state == TCPS_TIME_WAIT ?	"state-time-wait" :
	"<unknown>";

/*
 * tcpsinfo contains stable TCP details from tcp_t.
 */
typedef struct tcpsinfo {
	uintptr_t tcps_addr;
	int tcps_local;			/* is delivered locally, boolean */
	int tcps_active;		/* active open (from here), boolean */
	uint16_t tcps_lport;		/* local port */
	uint16_t tcps_rport;		/* remote port */
	string tcps_laddr;		/* local address, as a string */
	string tcps_raddr;		/* remote address, as a string */
	int32_t tcps_state;		/* TCP state */
	uint32_t tcps_iss;		/* Initial sequence # sent */
	uint32_t tcps_irs;		/* Initial sequence # received */
	uint32_t tcps_suna;		/* sequence # sent but unacked */
	uint32_t tcps_smax;		/* highest sequence number sent */
	uint32_t tcps_snxt;		/* next sequence # to send */
	uint32_t tcps_rack;		/* sequence # we have acked */
	uint32_t tcps_rnxt;		/* next sequence # expected */
	u_long tcps_swnd;		/* send window size */
	int32_t tcps_snd_ws;		/* send window scaling */
	uint32_t tcps_swl1;		/* window update seg seq number */
	uint32_t tcps_swl2;		/* window update seg ack number */
	uint32_t tcps_rup;		/* receive urgent pointer */
	uint32_t tcps_radv;		/* advertised window */
	u_long tcps_rwnd;		/* receive window size */
	int32_t tcps_rcv_ws;		/* receive window scaling */
	u_long tcps_cwnd;		/* congestion window */
	u_long tcps_cwnd_ssthresh;	/* threshold for congestion avoidance */
	uint32_t tcps_srecover;		/* for use in NewReno Fast Recovery */
	uint32_t tcps_sack_fack;	/* SACK sequence # we have acked */
	uint32_t tcps_sack_snxt;	/* next SACK seq # for retransmission */
	uint32_t tcps_rto;		/* round-trip timeout, msec */
	uint32_t tcps_mss;		/* max segment size */
	int tcps_retransmit;		/* retransmit send event, boolean */
	int tcps_srtt;			/* smoothed RTT in units of (TCP_RTT_SCALE*hz) */
	int tcps_debug;			/* socket has SO_DEBUG set */
	int tcps_cookie;		/* expose the socket's SO_USER_COOKIE */
	int32_t tcps_dupacks;		/* consecutive dup acks received */
	uint32_t tcps_rtttime;		/* RTT measurement start time */
	uint32_t tcps_rtseq;		/* sequence # being timed */
	uint32_t tcps_ts_recent;	/* timestamp echo data */
} tcpsinfo_t;

/*
 * tcplsinfo provides the old tcp state for state changes.
 */
typedef struct tcplsinfo {
	int32_t tcps_state;		/* previous TCP state */
} tcplsinfo_t;

/*
 * tcpinfo is the TCP header fields.
 */
typedef struct tcpinfo {
	uint16_t tcp_sport;		/* source port */
	uint16_t tcp_dport;		/* destination port */
	uint32_t tcp_seq;		/* sequence number */
	uint32_t tcp_ack;		/* acknowledgment number */
	uint8_t tcp_offset;		/* data offset, in bytes */
	uint8_t tcp_flags;		/* flags */
	uint16_t tcp_window;		/* window size */
	uint16_t tcp_checksum;		/* checksum */
	uint16_t tcp_urgent;		/* urgent data pointer */
	struct tcphdr *tcp_hdr;		/* raw TCP header */
} tcpinfo_t;

/*
 * A clone of tcpinfo_t used to handle the fact that the TCP input path
 * overwrites some fields of the TCP header with their host-order equivalents.
 * Unfortunately, DTrace doesn't let us simply typedef a new name for struct
 * tcpinfo and define a separate translator for it.
 */
typedef struct tcpinfoh {
	uint16_t tcp_sport;		/* source port */
	uint16_t tcp_dport;		/* destination port */
	uint32_t tcp_seq;		/* sequence number */
	uint32_t tcp_ack;		/* acknowledgment number */
	uint8_t tcp_offset;		/* data offset, in bytes */
	uint8_t tcp_flags;		/* flags */
	uint16_t tcp_window;		/* window size */
	uint16_t tcp_checksum;		/* checksum */
	uint16_t tcp_urgent;		/* urgent data pointer */
	struct tcphdr *tcp_hdr;		/* raw TCP header */
} tcpinfoh_t;

#pragma D binding "1.6.3" translator
translator csinfo_t < struct tcpcb *p > {
	cs_addr =	NULL;
	cs_cid =	(uint64_t)(p == NULL ? 0 : p->t_inpcb);
	cs_pid =	0;
	cs_zoneid =	0;
};

#pragma D binding "1.6.3" translator
translator tcpsinfo_t < struct tcpcb *p > {
	tcps_addr =		(uintptr_t)p;
	tcps_local =		-1; /* XXX */
	tcps_active =		-1; /* XXX */
	tcps_lport =		p == NULL ? 0 : ntohs(p->t_inpcb->inp_inc.inc_ie.ie_lport);
	tcps_rport =		p == NULL ? 0 : ntohs(p->t_inpcb->inp_inc.inc_ie.ie_fport);
	tcps_laddr =		p == NULL ? "<unknown>" :
	    p->t_inpcb->inp_vflag == INP_IPV4 ?
	    inet_ntoa(&p->t_inpcb->inp_inc.inc_ie.ie_dependladdr.id46_addr.ia46_addr4.s_addr) :
	    inet_ntoa6(&p->t_inpcb->inp_inc.inc_ie.ie_dependladdr.id6_addr);
	tcps_raddr =		p == NULL ? "<unknown>" :
	    p->t_inpcb->inp_vflag == INP_IPV4 ?
	    inet_ntoa(&p->t_inpcb->inp_inc.inc_ie.ie_dependfaddr.id46_addr.ia46_addr4.s_addr) :
	    inet_ntoa6(&p->t_inpcb->inp_inc.inc_ie.ie_dependfaddr.id6_addr);
	tcps_state =		p == NULL ? -1 : p->t_state;
	tcps_iss =		p == NULL ? 0  : p->iss;
	tcps_irs =		p == NULL ? 0  : p->irs;
	tcps_suna =		p == NULL ? 0  : p->snd_una;
	tcps_smax =		p == NULL ? 0  : p->snd_max;
	tcps_snxt =		p == NULL ? 0  : p->snd_nxt;
	tcps_rack =		p == NULL ? 0  : p->last_ack_sent;
	tcps_rnxt =		p == NULL ? 0  : p->rcv_nxt;
	tcps_swnd =		p == NULL ? -1 : p->snd_wnd;
	tcps_snd_ws =		p == NULL ? -1 : p->snd_scale;
	tcps_swl1 =		p == NULL ? -1 : p->snd_wl1;
	tcps_swl2 = 		p == NULL ? -1 : p->snd_wl2;
	tcps_radv =		p == NULL ? -1 : p->rcv_adv;
	tcps_rwnd =		p == NULL ? -1 : p->rcv_wnd;
	tcps_rup =		p == NULL ? -1 : p->rcv_up;
	tcps_rcv_ws =		p == NULL ? -1 : p->rcv_scale;
	tcps_cwnd =		p == NULL ? -1 : p->snd_cwnd;
	tcps_cwnd_ssthresh =	p == NULL ? -1 : p->snd_ssthresh;
	tcps_srecover =		p == NULL ? -1 : p->snd_recover;
	tcps_sack_fack =	p == NULL ? 0  : p->snd_fack;
	tcps_sack_snxt =	p == NULL ? 0  : p->sack_newdata;
	tcps_rto =		p == NULL ? -1 : (p->t_rxtcur * 1000) / `hz;
	tcps_mss =		p == NULL ? -1 : p->t_maxseg;
	tcps_retransmit =	p == NULL ? -1 : p->t_rxtshift > 0 ? 1 : 0;
	tcps_srtt =		p == NULL ? -1 : p->t_srtt;   /* smoothed RTT in units of (TCP_RTT_SCALE*hz) */
	tcps_debug =		p == NULL ? 0 :
	    p->t_inpcb->inp_socket->so_options & 1;
	tcps_cookie =		p == NULL ? -1 :
	    p->t_inpcb->inp_socket->so_user_cookie;
	tcps_dupacks =		p == NULL ? -1 : p->t_dupacks;
	tcps_rtttime =		p == NULL ? -1 : p->t_rtttime;
	tcps_rtseq =		p == NULL ? -1 : p->t_rtseq;
	tcps_ts_recent =	p == NULL ? -1 : p->ts_recent;
};

#pragma D binding "1.6.3" translator
translator tcpinfo_t < struct tcphdr *p > {
	tcp_sport =	p == NULL ? 0  : ntohs(p->th_sport);
	tcp_dport =	p == NULL ? 0  : ntohs(p->th_dport);
	tcp_seq =	p == NULL ? -1 : ntohl(p->th_seq);
	tcp_ack =	p == NULL ? -1 : ntohl(p->th_ack);
	tcp_offset =	p == NULL ? -1 : (p->th_off >> 2);
	tcp_flags =	p == NULL ? 0  : p->th_flags;
	tcp_window =	p == NULL ? 0  : ntohs(p->th_win);
	tcp_checksum =	p == NULL ? 0  : ntohs(p->th_sum);
	tcp_urgent =	p == NULL ? 0  : ntohs(p->th_urp);
	tcp_hdr =	(struct tcphdr *)p;
};

/*
 * This translator differs from the one for tcpinfo_t in that the sequence
 * number, acknowledgement number, window size and urgent pointer are already
 * in host order and thus don't need to be converted.
 */
#pragma D binding "1.6.3" translator
translator tcpinfoh_t < struct tcphdr *p > {
	tcp_sport =	p == NULL ? 0  : ntohs(p->th_sport);
	tcp_dport =	p == NULL ? 0  : ntohs(p->th_dport);
	tcp_seq =	p == NULL ? -1 : p->th_seq;
	tcp_ack =	p == NULL ? -1 : p->th_ack;
	tcp_offset =	p == NULL ? -1 : (p->th_off >> 2);
	tcp_flags =	p == NULL ? 0  : p->th_flags;
	tcp_window =	p == NULL ? 0  : p->th_win;
	tcp_checksum =	p == NULL ? 0  : ntohs(p->th_sum);
	tcp_urgent =	p == NULL ? 0  : p->th_urp;
	tcp_hdr =	(struct tcphdr *)p;
};

#pragma D binding "1.6.3" translator
translator tcplsinfo_t < int s > {
	tcps_state =	s;
};


/* Support for TCP debug */

#pragma D binding "1.12.1" TA_INPUT
inline int TA_INPUT =	0;
#pragma D binding "1.12.1" TA_OUTPUT
inline int TA_OUTPUT =	1;
#pragma D binding "1.12.1" TA_USER
inline int TA_USER =	2;
#pragma D binding "1.12.1" TA_RESPOND
inline int TA_RESPOND =	3;
#pragma D binding "1.12.1" TA_DROP
inline int TA_DROP =	4;

/* direction strings. */

#pragma D binding "1.12.1" tcpdebug_dir_string
inline string tcpdebug_dir_string[uint8_t direction] =
	direction == TA_INPUT ?	"input" :
	direction == TA_OUTPUT ? "output" :
	direction == TA_USER ? "user" :
	direction == TA_RESPOND ? "respond" :
	direction == TA_OUTPUT ? "drop" :
	"unknown" ;

#pragma D binding "1.12.1" tcpflag_string
inline string tcpflag_string[uint8_t flags] =
	flags & TH_FIN ?	"FIN" :
	flags & TH_SYN ?	"SYN" :
	flags & TH_RST ?	"RST" :
	flags & TH_PUSH ?	"PUSH" :
	flags & TH_ACK ?	"ACK" :
	flags & TH_URG ?	"URG" :
	flags & TH_ECE ?	"ECE" :
	flags & TH_CWR ?	"CWR" :
	"unknown" ;

#pragma D binding "1.12.1" PRU_ATTACH
inline int PRU_ATTACH		= 0;
#pragma D binding "1.12.1" PRU_DETACH
inline int PRU_DETACH		= 1;
#pragma D binding "1.12.1" PRU_BIND
inline int PRU_BIND		= 2;
#pragma D binding "1.12.1" PRU_LISTEN
inline int PRU_LISTEN		= 3;
#pragma D binding "1.12.1" PRU_CONNECT
inline int PRU_CONNECT		= 4;
#pragma D binding "1.12.1" PRU_ACCEPT
inline int PRU_ACCEPT		= 5 ;
#pragma D binding "1.12.1" PRU_DISCONNECT
inline int PRU_DISCONNECT	= 6;
#pragma D binding "1.12.1" PRU_SHUTDOWN
inline int PRU_SHUTDOWN		= 7;
#pragma D binding "1.12.1" PRU_RCVD
inline int PRU_RCVD		= 8;
#pragma D binding "1.12.1" PRU_SEND
inline int PRU_SEND		= 9;
#pragma D binding "1.12.1" PRU_ABORT
inline int PRU_ABORT		= 10;
#pragma D binding "1.12.1" PRU_CONTROL
inline int PRU_CONTROL		= 11;
#pragma D binding "1.12.1" PRU_SENSE
inline int PRU_SENSE		= 12;
#pragma D binding "1.12.1" PRU_RCVOOB
inline int PRU_RCVOOB		= 13;
#pragma D binding "1.12.1" PRU_SENDOOB
inline int PRU_SENDOOB		= 14;
#pragma D binding "1.12.1" PRU_SOCKADDR
inline int PRU_SOCKADDR		= 15;
#pragma D binding "1.12.1" PRU_PEERADDR
inline int PRU_PEERADDR		= 16;
#pragma D binding "1.12.1" PRU_CONNECT2
inline int PRU_CONNECT2		= 17;
#pragma D binding "1.12.1" PRU_FASTTIMO
inline int PRU_FASTTIMO		= 18;
#pragma D binding "1.12.1" PRU_SLOWTIMO
inline int PRU_SLOWTIMO		= 19;
#pragma D binding "1.12.1" PRU_PROTORCV
inline int PRU_PROTORCV		= 20;
#pragma D binding "1.12.1" PRU_PROTOSEND
inline int PRU_PROTOSEND	= 21;
#pragma D binding "1.12.1" PRU_SEND_EOF
inline int PRU_SEND_EOF		= 22;
#pragma D binding "1.12.1" PRU_SOSETLABEL
inline int PRU_SOSETLABEL	= 23;
#pragma D binding "1.12.1" PRU_CLOSE
inline int PRU_CLOSE		= 24;
#pragma D binding "1.12.1" PRU_FLUSH
inline int PRU_FLUSH		= 25;

#pragma D binding "1.12.1" prureq_string
inline string prureq_string[uint8_t req] =
	req == PRU_ATTACH ?	"ATTACH" :
	req == PRU_DETACH ?	"DETACH" :
	req == PRU_BIND ?	"BIND" :
	req == PRU_LISTEN ?	"LISTEN" :
	req == PRU_CONNECT ?	"CONNECT" :
	req == PRU_ACCEPT ?	"ACCEPT" :
	req == PRU_DISCONNECT ?	"DISCONNECT" :
	req == PRU_SHUTDOWN ?	"SHUTDOWN" :
	req == PRU_RCVD ?	"RCVD" :
	req == PRU_SEND ?	"SEND" :
	req == PRU_ABORT ?	"ABORT" :
	req == PRU_CONTROL ?	"CONTROL" :
	req == PRU_SENSE ?	"SENSE" :
	req == PRU_RCVOOB ?	"RCVOOB" :
	req == PRU_SENDOOB ?	"SENDOOB" :
	req == PRU_SOCKADDR ?	"SOCKADDR" :
	req == PRU_PEERADDR ?	"PEERADDR" :
	req == PRU_CONNECT2 ?	"CONNECT2" :
	req == PRU_FASTTIMO ?	"FASTTIMO" :
	req == PRU_SLOWTIMO ?	"SLOWTIMO" :
	req == PRU_PROTORCV ?	"PROTORCV" :
	req == PRU_PROTOSEND ?	"PROTOSEND" :
	req == PRU_SEND ?	"SEND_EOF" :
	req == PRU_SOSETLABEL ?	"SOSETLABEL" :
	req == PRU_CLOSE ?	"CLOSE" :
	req == PRU_FLUSH ?	"FLUSE" :
	"unknown" ;
